update_npm_dependencies.rb

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env ruby
+
+require 'optparse'
+require 'json'
+
+def package_name(npm_module_name)
+  "nodejs-#{npm_module_name}"
+end
+
+def package_dir(npm_module_name)
+  "packages/foreman/#{package_name(npm_module_name)}"
+end
+
+def package_exist?(npm_module_name)
+  File.directory?(package_dir(npm_module_name))
+end
+
+def current_version(npm_module_name)
+  if package_exist?(npm_module_name)
+    spec_file = "#{package_dir(npm_module_name)}/#{package_name(npm_module_name)}.spec"
+    version_line = File.readlines(spec_file).grep(/^\W*Version:.*/).first
+    version_line.gsub(/^\W*Version:\W*/, '').strip
+  end
+end
+
+def bundle_strategy(npm_module_name)
+  if package_exist?(npm_module_name)
+    # new packages default to bundle
+    'bundle'
+  else
+    dir = package_dir(npm_module_name)
+    zipped_dependencies = Dir["#{dir}/*.tgz"]
+    if zipped_dependencies.length > 1
+      'bundle'
+    else
+      'single'
+    end
+  end
+end
+
+def detect_action(npm_module_name, details)
+  if package_exist?(npm_module_name)
+    if details['version'] == current_version(npm_module_name)
+      :skip
+    else
+      :update
+    end
+  else
+    :add
+  end
+end
+
+def add_package_cmd(npm_module_name, module_details)
+  strategy = bundle_strategy(npm_module_name)
+  cmd = ''
+  if @options[:git_packager]
+    cmd += 'RPM_PACKAGER="$(git config user.name) <$(git config user.email)>" '
+  end
+  cmd += "./add_npm_package.sh #{npm_module_name} #{module_details['version']} #{strategy}"
+  cmd
+end
+
+
+@options = {}
+opts = OptionParser.new do |opts|
+  opts.banner = [
+    "Usage: #{__FILE__} [options]",
+    "",
+    "Updates npm packages based on json output of npm list.",
+    "Make sure you generate the input json file without development and optional dependencies:",
+    "    npm list --production --no-optional --json",
+    "",
+    ""
+  ].join("\n")
+
+  opts.separator('Options:')
+  opts.on('--package-list FILE', 'Path to a json file with output of npm list (required)') do |val|
+    if !File.exist?(val)
+      puts "File #{val} doesn't exist."
+      exit 1
+    end
+    @options[:package_list_file] = val
+  end
+
+  opts.on('--only NPM_MODULES', 'Comma separated list of npm modules to update, others are skipped') do |val|
+    @options[:only] = val.split(',')
+  end
+
+  opts.on('--except NPM_MODULES', 'Comma separated list of npm modules to be excluded from updating') do |val|
```

Comment:
In `update-requirements` we have a blacklist. I wonder if we could unify that somewhere in a config to prevent duplication.

Reply:
I intended the options `--only` and `--except` to be used rather for fixing issues when some packages fail or for some other exceptional cases so I'd like to keep them in regardless of `update-requirements`.

The comment in `update-requirements` says:
```
# We need a blacklist of packages that are devDependencies but not needed to build
```
Is it only dev dependencies that we're filtering now? If it's so we can sort it out with `npm list --production` that lists the modules without dev dependencies already.

Reply:
I found it tricky. I opted to skip some dependencies that were only used for tests like jest. It's entirely possible I took too much of a shortcut in some places. Sadly I don't think there's a testDependencies to split it out further or bundler groups that give you complete freedom.

Reply:
I'm not sure I understand. Pardon my ignorance, I probably lack some background knowledge.
So you think using `npm list --production` is sufficient or do we still need the blacklist?

If we do need the blacklist, what's the key to decide if a package should go in or be put on the blacklist?